src/main/telemetry/msp_shared.c

```diff
@@ -0,0 +1,217 @@
+    sbufSwitchToReader(&mspPackage.responsePacket->buf, mspPackage.responseBuffer);
+}
+
+bool handleMspFrame(uint8_t *frameStart, uint8_t *frameEnd)
+{
+    static uint8_t mspStarted = 0;
+    static uint8_t lastSeq = 0;
+
+    mspPacket_t *packet = mspPackage.requestPacket;
+    sbuf_t *frameBuf = &mspPackage.requestFrame;
+
+    if (sbufBytesRemaining(&mspPackage.responsePacket->buf) > 0) {
+        return false;
+    } 
+      
+    if (mspStarted == 0) {
+        initSharedMsp();
+    }
+    
+    uint8_t frameLength = frameEnd - frameStart;
+    uint8_t frame[frameLength];
+
+    memcpy(&frame, frameStart, frameLength);
+
+    frameBuf->ptr = (uint8_t*)&frame;
+    frameBuf->end = (uint8_t*)&frame + frameLength;
+
+    uint8_t header = sbufReadU8(frameBuf);
+    uint8_t seqNumber = header & TELEMETRY_MSP_SEQ_MASK; 
+    uint8_t version = (header & TELEMETRY_MSP_VER_MASK) >> TELEMETRY_MSP_VER_SHIFT;
+
+    if (version != TELEMETRY_MSP_VERSION) {
+        sendMspErrorResponse(TELEMETRY_MSP_VER_MISMATCH, 0);
+        return true;
+    }
+
+    if (header & TELEMETRY_MSP_START_FLAG) {
+        // first packet in sequence
+        uint8_t mspPayloadSize = sbufReadU8(frameBuf);
+        
+        packet->cmd = sbufReadU8(frameBuf);
+        packet->result = 0;
+        packet->buf.ptr = mspPackage.requestBuffer;
+        packet->buf.end = mspPackage.requestBuffer + mspPayloadSize;
+
+        checksum = mspPayloadSize ^ packet->cmd;
+        mspStarted = 1;
+    } else if (!mspStarted) {
+        // no start packet yet, throw this one away
+        return false;
+    } else if (((lastSeq + 1) & TELEMETRY_MSP_SEQ_MASK) != seqNumber) {
+        // packet loss detected!
+        mspStarted = 0;
+        return false; 
+    }
+
+    while ((frameBuf->ptr < frameBuf->end) && sbufBytesRemaining(&packet->buf)) {
+        checksum ^= *frameBuf->ptr;
+        sbufWriteU8(&packet->buf, sbufReadU8(frameBuf));
+    }
+
+    if (sbufBytesRemaining(frameBuf) == 0) {
+        lastSeq = seqNumber;
+        return false;
+    }
+
+    if (checksum != *frameBuf->ptr) {
+        mspStarted = 0;
+        sendMspErrorResponse(TELEMETRY_MSP_CRC_ERROR, packet->cmd);
+        return true;
+    } 
+
+    mspStarted = 0;
+    sbufSwitchToReader(&packet->buf, mspPackage.requestBuffer);
+    processMspPacket();
+    return true;
+}
+
+bool sendMspReply(uint8_t payloadSize, mspResponseFnPtr responseFn)
+{
+    static uint8_t checksum = 0;
+    static uint8_t seq = 0;
+
+    uint8_t packet[payloadSize];
+    uint8_t *p = packet;
+    uint8_t *end = p + payloadSize;
+
+    sbuf_t *txBuf = &mspPackage.responsePacket->buf;
+
+    // detect first reply packet
+    if (txBuf->ptr == mspPackage.responseBuffer) {
+
+        // header
+        uint8_t head = TELEMETRY_MSP_START_FLAG | (seq++ & TELEMETRY_MSP_SEQ_MASK);
+        if (mspPackage.responsePacket->result < 0) {
+            head |= TELEMETRY_MSP_ERROR_FLAG;
+        }
+        *p++ = head;
```

Comment:
Can you create an `sbuf` over the packet and use sbuf functions rather than direct pointer arithmetic.

Reply:
thats sounds good :) we wanted to change existing code as less as possible, so smartport msp doesnt get affected by this changes but cleanups should be happen :)

Reply:
Sure. was planning to do this anyway :)